File: packages/code-review/src/finding-ids.ts
/**
 * IDs estáveis de achados de review (tema, não path/linha).
 */

const FINDING_THEME = /^[^:\n]+:\d+(?:-\d+)?\s*[—\-]\s*(.+)$/;

export function slugify(text: string): string {
  const base = String(text)
    .toLowerCase()
    .replace(/[^\w\s-]/g, "")
    .replace(/[-\s]+/g, "-")
    .replace(/^-+|-+$/g, "");
  return base.slice(0, 80) || "finding";
}

/** Remove prefixo arquivo:linha — do título do achado. */
export function extractFindingTheme(text: string): string {
  const stripped = String(text).trim();
  if (!stripped) return "finding";
  const match = FINDING_THEME.exec(stripped);
  if (match) return match[1].trim();
  return stripped;
}

/** ID estável entre PRs — só a descrição do achado, não path/linha. */
export function stableFindingId(text: string): string {
  return slugify(extractFindingTheme(text));
}

export { FINDING_THEME };

File: packages/code-review/tools/lib/finding_ids.py
"""IDs estáveis de achados de review (tema, não path/linha)."""
from __future__ import annotations

import re

FINDING_THEME = re.compile(r"^[^:\n]+:\d+(?:-\d+)?\s*[—\-]\s*(.+)$")


def slugify(text: str) -> str:
    base = re.sub(r"[^\w\s-]", "", text.lower())
    base = re.sub(r"[-\s]+", "-", base).strip("-")
    return base[:80] or "finding"


def extract_finding_theme(text: str) -> str:
    """Remove prefixo arquivo:linha — do título do achado."""
    stripped = text.strip()
    if not stripped:
        return "finding"
    match = FINDING_THEME.match(stripped)
    if match:
        return match.group(1).strip()
    return stripped


def stable_finding_id(text: str) -> str:
    """ID estável entre PRs — só a descrição do achado, não path/linha."""
    return slugify(extract_finding_theme(text))
